pulp-glue/pulp_glue/common/openapi.py

```diff
@@ -53,36 +55,79 @@ class AuthProviderBase:
     Returned auth objects need to be compatible with `requests.auth.AuthBase`.
     """
 
-    def basic_auth(self) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
-        """Implement this to provide means of http basic auth."""
+    auth_type: str = ""
+
+    def auth(self, request: requests.PreparedRequest) -> requests.auth.AuthBase:
+        """Implement this to provide means of http auth."""
         return None
 
-    def __call__(
-        self,
-        security: t.List[t.Dict[str, t.List[str]]],
-        security_schemes: t.Dict[str, t.Dict[str, t.Any]],
-    ) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
-        for proposal in security:
-            if [security_schemes[name] for name in proposal] == [
-                {"type": "http", "scheme": "basic"}
-            ]:
-                result = self.basic_auth()
-                if result:
-                    return result
-        raise OpenAPIError(_("No suitable auth scheme found."))
+    def __call__(self, request: requests.PreparedRequest) -> requests.PreparedRequest:
+        return self.auth(request)
+
+    # def __call__(
+    #     self,
+    #     security: t.List[t.Dict[str, t.List[str]]],
+    #     security_schemes: t.Dict[str, t.Dict[str, t.Any]],
+    # ) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
+    #     for proposal in security:
+    #         if [security_schemes[name] for name in proposal] == [
+    #             {"type": "http", "scheme": "basic"}
+    #         ]:
+    #             result = self.basic_auth()
+    #             if result:
+    #                 return result
+    #     raise OpenAPIError(_("No suitable auth scheme found."))
 
 
 class BasicAuthProvider(AuthProviderBase):
     """
     Reference Implementation for AuthProviderBase providing basic auth with `username`, `password`.
     """
 
+    auth_type = "basic"
+
     def __init__(self, username: str, password: str):
         self.username = username
         self.password = password
 
-    def basic_auth(self) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
-        return (self.username, self.password)
+    def auth(self, request) -> requests.PreparedRequest:
+        return requests.auth.HTTPBasicAuth(self.username, self.password)(request)
+
+
+class TokenAuthProvider(AuthProviderBase):
+    """
+    Reference implementation for a Token based authentication.
+    """
+
+    auth_type = "token"
+
+    def __init__(self, token: str):
+        self.token = token
+
+    def auth(self, request) -> requests.PreparedRequest:
+        request.headers["Authorization"] = self.token
+        return request
+
+
+class ThirdPartyAuthProvider(AuthProviderBase):
+
+    auth_type = "third_party"
```

Comment:
Why can we not just call it oauth2?
Also isn't there a requests-oauth out there?

Reply:
Renamed it, and yes, we're using `request-oauthlib`. Here https://github.com/pulp/pulp-cli/blob/0738418f7247b7e901c03ea69b6997cb6842653e/pulp-glue/pyproject.toml#L28